Update Customer Effort Score Modal

adrianduffell opened the pull request:
This PR updates the placeholder modal with a new `Feedback` component containing [WordPress package components](https://developer.wordpress.org/block-editor/packages/packages-components/). 

The `RadioControl` component has been customized to display emoji buttons for the user to provide a score.

When a score of 1 or 2 is given, an optional comments field is provided.

An error message appears if the feedback is submitted without a score.

A basic small screen layout is in place. It is expected this will be refined further in later updates.

### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

#### Small Screen
![Mobile](https://user-images.githubusercontent.com/9312929/97831124-1a1e3d80-1d0a-11eb-8bdb-20efcaeee458.png)

#### Desktop Screen
![Desktop](https://user-images.githubusercontent.com/9312929/97831235-5fdb0600-1d0a-11eb-8647-f59bd08517bd.png)

### Detailed test instructions:

- Delete any previous WP options for Customer Effort Score, e.g. `$ wp option delete woocommerce_ces_shown_for_actions`
- Update an existing product.
- On the next page, wait for the notification and press Give Feedback.

<img width="436" alt="Screen Shot 2020-11-18 at 2 13 50 pm" src="https://user-images.githubusercontent.com/9312929/99492380-6cb55600-29a8-11eb-93b2-12df1e819dfe.png">

- Confirm feedback modal appears
- Confirm clicking on send button validates score and displays error message
- Confirm hovering over emoji button displays hover state with blue background
- Confirm clicking emoji button displays active state with blue background
- Confirm comments field displays for the "Very Difficult" and "Somewhat Difficult" scores
- Resize window to small size and confirm modal is responsive.
- Close modal window

adrianduffell committed:
Add Customer Effort Score modal

adrianduffell committed:
Add style for selected emoji

adrianduffell committed:
Update modal layout to match design

adrianduffell committed:
Update colors to use color studio variables

adrianduffell committed:
Refactor SCSS to reduce nesting

adrianduffell commented:
I have some tests planned for this but looking to implement those for #5336.

mattsherman commented on packages/customer-effort-score/src/index.js:
```diff
@@ -27,22 +33,53 @@ function CustomerEffortScore( {
+			{ isOpen && (
+				<Modal
+					className="woocommerce-customer-effort-score"
+					title="Please share your feedback"
```
Text needs to be wrapped for translation. Use

```
import { __ } from '@wordpress/i18n';
```

adrianduffell replied:
This is now fixed in 111472a652fc4693871aa76be21efe348b5a70fe.

mattsherman commented on packages/customer-effort-score/src/index.js:
```diff
@@ -27,22 +33,53 @@ function CustomerEffortScore( {
+					<RadioControl
+						selected={ score }
+						options={ [
+							{ label: 'Very difficult', value: '1' },
+							{ label: 'Somewhat difficult', value: '2' },
+							{ label: 'Neutral', value: '3' },
+							{ label: 'Somewhat easy', value: '4' },
+							{ label: 'Very easy', value: '5' },
```
Text needs to be wrapped for translation. Use

```
import { __ } from '@wordpress/i18n';
```

adrianduffell replied:
This is now fixed in 111472a652fc4693871aa76be21efe348b5a70fe.

mattsherman commented on packages/customer-effort-score/src/index.js:
```diff
@@ -27,22 +33,53 @@ function CustomerEffortScore( {
+
+					<div className="woocommerce-customer-effort-score__buttons">
+						<Button isTertiary onClick={ closeModal }>
+							Cancel
```
Text needs to be wrapped for translation. Use

```
import { __ } from '@wordpress/i18n';
```

mattsherman commented on packages/customer-effort-score/src/index.js:
```diff
@@ -27,22 +33,53 @@ function CustomerEffortScore( {
+							Cancel
+						</Button>
+						<Button isPrimary onClick={ sendScore }>
+							Send
```
Text needs to be wrapped for translation. Use

```
import { __ } from '@wordpress/i18n';
```

adrianduffell replied:
This is now fixed in 111472a652fc4693871aa76be21efe348b5a70fe.

mattsherman reviewed (changes requested):
I didn't fully review the CSS yet, but wanted to pass on some initial comments (I'll give the CSS a more thorough review later!)...

Overall, great start to this!

Text will need to be wrapped for translation.

This implementation is not keyboard accessible (I didn't check screen reader).

becdetat commented:
I think it needs a bit of margin around the buttons to make the longer text wrap and help with text pushing right against the edges of the highlight, to make it match the designs more closely:

![image](https://user-images.githubusercontent.com/224531/97934880-b0159f00-1dc2-11eb-890b-b9e77cab01ef.png)

becdetat commented:
The mobile landscape view (iphone 5) should have horizontal buttons, I think having the breakpoint at 600px is too wide.
![image](https://user-images.githubusercontent.com/224531/97935033-20242500-1dc3-11eb-987d-307ed98229fa.png)

becdetat commented:
It looks like the content isn't centred in some resolutions:
![image](https://user-images.githubusercontent.com/224531/97935126-5eb9df80-1dc3-11eb-901e-4247adc164de.png)

becdetat commented:
This may be a problem with the mobile breakpoints for the modal itself, but the action buttons aren't spaced correctly:
![image](https://user-images.githubusercontent.com/224531/97935206-9c1e6d00-1dc3-11eb-9f8f-6a03dd1dd499.png)

adrianduffell committed:
Refactor render logic to reduce nesting

adrianduffell committed:
Add translation to strings

adrianduffell committed:
Add keyboard support to emoji buttons

adrianduffell committed:
Fix button misalignment

adrianduffell committed:
Fix missing white space around button text

adrianduffell commented:
Thanks @mattsherman !

> Text will need to be wrapped for translation.

Ah good catch. It's now in place using `@wordpress/i18n`.
 
> This implementation is not keyboard accessible (I didn't check screen reader).

Thanks for testing this out 👍  I've now put in place a more accessible technique to hide the underlying radio inputs, which has enabled keyboard access for the emoji buttons. Additionally I've added an outline around the focused button when using a keyboard.

![Screen Shot 2020-11-03 at 12 33 14 pm](https://user-images.githubusercontent.com/9312929/97950773-fa068100-1dd2-11eb-82a3-20406e1794d8.png)

Do you think any further work is needed for keyboard accessibility?

adrianduffell commented:
> The mobile landscape view (iphone 5) should have horizontal buttons, I think having the breakpoint at 600px is too wide.

Thanks for testing this @becdetat, I think it would look better for iPhone 5 in landscape to have the horizontal buttons too. It has a big enough size to display it like that. The problem is the next available breakpoint (400px) is too small to fit all of the horizontal buttons. See screenshots below.

That's assuming 400px is the next smallest breakpoint. I am looking at the list [here](https://github.com/woocommerce/woocommerce-admin/blob/7c1299c5e2b088d29d8158a3ad806672ef1d14ef/client/stylesheets/abstracts/_breakpoints.scss#L10).

Would it be worth making the buttons smaller for screens between 400-600px perhaps?

### Screenshots

![Screen Shot 2020-11-03 at 2 00 28 pm](https://user-images.githubusercontent.com/9312929/97953920-f5df6100-1ddc-11eb-846b-b0ad432fee91.png)

_iPhone 5_

![Screen Shot 2020-11-03 at 1 40 27 pm](https://user-images.githubusercontent.com/9312929/97953025-34275100-1dda-11eb-9d85-6868fd7906f2.png)

_401px screen_

adrianduffell commented:
Thank you @becdetat !

> I think it needs a bit of margin around the buttons to make the longer text wrap and help with text pushing right against the edges of the highlight

This is now fixed in 6703de90cbd471dac048896aff27bc58efb0a1a6. The `Somewhat difficult` text needs to wrap onto another line to match the design as well.

> It looks like the content isn't centred in some resolutions:

Do you still see this issue? I saw it earlier but cannot reproduce now 😕 

![Screen Shot 2020-11-03 at 2 09 16 pm](https://user-images.githubusercontent.com/9312929/97954580-dea17300-1dde-11eb-85ae-9806b12896f9.png)

becdetat commented:
Pinging @elizaan36 to check over the design :-)

Also @adrianduffell I just remembered, do the emojis work in IE11?

mattsherman commented:
@adrianduffell Thanks for wrapping the strings for translation and making the modal keyboard accessible! 🙌

I am still able to reproduce the non-centered options on an iPhone 5 sized display:

<img width="495" alt="Screen Shot 2020-11-03 at 15 47 26" src="https://user-images.githubusercontent.com/2098816/98040443-4a3f0c80-1dee-11eb-82ba-fb6a5a71e118.png">

The caption for the close modal button (the "X") is cut-off. I'm not sure if this is a general issue with the modal component or our use of it:

<img width="680" alt="Screen Shot 2020-11-03 at 15 45 01" src="https://user-images.githubusercontent.com/2098816/98040550-6f337f80-1dee-11eb-8faf-d61472957577.png">

On Safari 14, the height of the options differs, which feels a little less polished than ideal:

![Option heights](https://user-images.githubusercontent.com/2098816/98040847-fd0f6a80-1dee-11eb-9a8a-74d0fcc56f41.gif)

adrianduffell commented:
> The caption for the close modal button (the "X") is cut-off. I'm not sure if this is a general issue with the modal component or our use of it:

@mattsherman I think this is a general issue with the modal component. I can reproduce it in the [storybook](https://wordpress.github.io/gutenberg/?path=/story/components-modal--default) using Firefox.

![Screen Shot 2020-11-04 at 1 19 06 pm](https://user-images.githubusercontent.com/9312929/98072261-95195c80-1ea0-11eb-9815-041e254472f6.png)

I'll add an issue about it on the Gutenberg repo.

elizaan36 commented:
All the comments here look spot on, and thanks for getting ahead on the mobile view and the attention to detail. I think the 400px breakpoint works to transition to the vertical layout for the options, and maybe we could give them a min-width (or % width) so they vary based on the size of the view?

> On Safari 14, the height of the options differs, which feels a little less polished than ideal
Would it make sense to give the height a fixed value of 125px rather than 100%?

Here's a mockup for <400px breakpoint:
![mobile](https://user-images.githubusercontent.com/5121465/98124665-1bba4380-1e79-11eb-9d63-4e023ba5325c.png)

Also, not sure if this is for another PR but upon selecting `Very difficult` or `Somewhat difficult` the optional comments field is shown in the modal.

adrianduffell commented:
> do the emojis work in IE11?

Yes, but IE is only capable of rendering them with one color.

![Screen Shot 2020-11-06 at 7 52 10 am](https://user-images.githubusercontent.com/9312929/98330417-58cb3500-2035-11eb-8a5d-aef3654022b5.png)

adrianduffell committed:
Refactor emoji button layout with flexbox

Flexbox makes more sense to use than a grid since the buttons need only a single row or column layout.

adrianduffell committed:
Merge branch 'main' into add/customer-effort-score-modal

# Conflicts:
#	packages/customer-effort-score/src/index.js

adrianduffell committed:
Refactor merge conflict resolution

Tidies up the diff by
- Using naming convention of “visible” instead of “open”.
- Restoring previous import order.

adrianduffell commented:
> I think the 400px breakpoint works to transition to the vertical layout for the options, and maybe we could give them a min-width (or % width) so they vary based on the size of the view?

@elizaan36 This sounds good to me. I think we could iterate on it in a follow-up PR to help this one over the line.

> Would it make sense to give the height a fixed value of 125px rather than 100%?

Thanks, I tried this but it didn’t provide enough height if the user has increased their font-size for accessibility.

![Screen Shot 2020-11-09 at 1 04 17 pm](https://user-images.githubusercontent.com/9312929/98502227-75ad7580-228c-11eb-8e8f-10ed30ef67d7.png)

![Screen Shot 2020-11-09 at 1 00 11 pm](https://user-images.githubusercontent.com/9312929/98502234-7940fc80-228c-11eb-89b3-5c35f5ccc95c.png)

Instead, I have switched the CSS layout to use flexbox instead of a grid. This makes more sense since there only needs to be a single row or column at a time. I think this will resolve the previously seen layout issues 🤞

adrianduffell commented:
I've now merged `main` which brings the integration with the snackbar component.

elizaan36 commented:
> Instead, I have switched the CSS layout to use flexbox instead of a grid.

elizaan36 commented:
Maybe for the follow up PR, but a couple things on the modal:

1. Looks like on mobile the selections need to be centered with the modal actions appearing on the bottom right without the need to scroll.
<img width="409" alt="Screen Shot 2020-11-09 at 5 02 07 PM" src="https://user-images.githubusercontent.com/5121465/98606845-95a16100-22ad-11eb-97df-4fb546c5c331.png">

2. Upon selecting Very difficult or Somewhat difficult the optional comments field is shown in the modal.
(shared the design in the comment above and in Figma)

Are cool with using the emoji unicode?

adrianduffell commented:
@elizaan36 Which browser/device did you use in your last screenshot?

I originally had in mind implementing the optional comments field in another PR - but I think it makes more sense to bring it in here now. I'll work on that next.

I think we could keep the unicode emoji here for now, but work on any changes to it in a follow-up PR.

elizaan36 commented:
Oh I should have said, it was with Firefox on my Mac laptop.

adrianduffell committed:
Add comments field

adrianduffell committed:
Update initial score value to undefined

adrianduffell committed:
Refactor SASS selectors

elizaan36 commented:
Hey, just tested this and the modal is looking good to me. The comments field appears too🎉
Just one thing - the comments field should only appear when `Very difficult` or `Somewhat difficult` is selected. Currently it appears for `Very difficult` or `Very easy`.

adrianduffell committed:
Merge branch 'main' into add/customer-effort-score-modal

adrianduffell committed:
Update sendScore with comments data

adrianduffell committed:
Fix comments field to display on low scores only

adrianduffell committed:
Refactor by extracting modal to own component

adrianduffell committed:
Add input validation

adrianduffell committed:
Add tests

adrianduffell committed:
Fix variable name

adrianduffell committed:
Update Feedback component documentation

adrianduffell committed:
Fix uneven button height

adrianduffell committed:
Merge branch 'main' into add/customer-effort-score-modal

# Conflicts:
#	packages/customer-effort-score/src/style.scss

adrianduffell committed:
Restore previous prop names

This ensures naming consistency with the CustomerEffortScore component.

adrianduffell committed:
Merge branch 'main' into add/customer-effort-score-modal

# Conflicts:
#	packages/customer-effort-score/src/index.js

adrianduffell committed:
Update variable name for consistency

adrianduffell commented:
This is ready for re-review. 

Since the last time, I have added the comments field, form validation, and tests. Testing instructions have been updated.

mattsherman committed:
Don't close modal if click outside it

mattsherman reviewed (changes requested):
With the commit I pushed, this works well.

Other than the suggestion to change the name of the `Feedback` component, this PR seems good to go. I think that any further cleanup could happen in follow-up PRs. It would be great to have this merged so that we can start getting a feel for how things will work, and work on follow-up PRs (like storing the comments).

mattsherman commented:
Nice work on this @adrianduffell !

elizaan36 commented on packages/customer-effort-score/src/feedback/index.js:
```diff
@@ -0,0 +1,134 @@
+	const sendScore = () => {
+		if ( ! [ '1', '2', '3', '4', '5' ].includes( score ) ) {
+			setErrorMessage(
+				__( 'Error: Feedback not provided.', 'woocommerce-admin' )
```
Or "This field is required" could work too.

adrianduffell replied:
Sounds good to me. Now updated in f47d7e8235bf4f3c9a380354994a54571f3a5142.

elizaan36 commented on packages/customer-effort-score/src/feedback/index.js:
```diff
@@ -0,0 +1,134 @@
+			{ ( score === '1' || score === '2' ) && (
+				<div className="woocommerce-customer-effort-score__comments">
+					<TextareaControl
+						label="Comments (Optional)"
```
I was going back and forth on that in the design, but I think capitalizing "Optional" feels right.

adrianduffell replied:
I think let's keep it capitalized then 🙂

elizaan36 commented on packages/customer-effort-score/src/feedback/index.js:
```diff
@@ -0,0 +1,135 @@
+			className="woocommerce-customer-effort-score"
+			title={ __( 'Please share your feedback', 'woocommerce-admin' ) }
+			onRequestClose={ closeModal }
+			shouldCloseOnClickOutside={ false }
```
I'm a little wary of this in the edge case where the actions to close the modal are hidden in some way and the user is stuck with the screen disabled. Do you think there's another way to solve this?

adrianduffell replied:
Good points raised here 👍 

Perhaps we could try to fix the resize bug upstream in the Gutenberg component instead.

I can reproduce it in the [component Storybook](https://wordpress.github.io/gutenberg/?path=/story/components-modal--default). It seems to close when the window size hits a responsive breakpoint.

mattsherman replied:
I agree that we should look into reporting/fixing the resize bug in upstream Gutenberg -- that clearly seems broken.

And, I also feel that having this modal dismissible by clicking outside of it is not ideal. I think the change to prevent this is a good one to make. 

Having a modal like this be so easily dismissible accidentally, where we are asking the user to provide input and there is no way for the user to easily bring back up the modal, does not feel user-friendly. Users mis-click/tap.

I don't think it is too likely that both the Cancel button at the bottom of the modal, and the X at the top, are somehow inaccessible. If they are, then there are some other issues we likely should be addressing. And, even if that were to happen, I think that 99% of users would just refresh the browser page, which would remove the dialog (and, nothing would be lost, since they only get this dialog after successfully saving something).

adrianduffell replied:
@mattsherman  I agree this is a good approach to take.

I do have some doubts about it since I think some users will try to click outside with an aim to close the modal quickly.

However, I feel like guarding against mis-clicks is more important taking into account there isn't a way for the user to recover and re-open the modal.

elizaan36 replied:
Yeah, I think this solution works to prevent closing by accident. And good call to report the resize bug in Gutenberg so hopefully we don't run into the same issue with future modals.

adrianduffell committed:
Rename Feedback component to CustomerFeedbackModal

adrianduffell committed:
Update error message text

adrianduffell committed:
Merge branch 'main' into add/customer-effort-score-modal

adrianduffell commented on packages/customer-effort-score/src/feedback/index.js:
```diff
@@ -0,0 +1,134 @@
+ * @param {Function} props.recordScoreCallback Function to call when the results are sent.
+ * @param {string} props.label                 Question to ask the customer.
+ */
+function Feedback( { recordScoreCallback, label } ) {
```
Thanks, I agree it does seem too generic. I have changed it to `CustomerFeedbackModal` in 10e4aac89afc6d97b60d9c92bf7450576954de4f. 

This is a close match to the existing doc comment for the component:

 “Provides a modal requesting customer feedback.”

adrianduffell committed:
Fix import name

adrianduffell committed:
Fix formatting

adrianduffell committed:
Update test to use mockRecordScoreCallback prop name

adrianduffell commented:
Thanks @mattsherman / @elizaan36 I've now addressed your feedback in the review comments.

mattsherman reviewed (approved):
This is feeling good. :shipit: 

Great job on this!

becdetat commented:
This is looking the 💣  @adrianduffell . I just have a few of pieces of feedback:

- The error message ("Please provide feedback by selecting an option above") should be cleared when an option is selected
- Can the options be numeric instead of strings? If this isn't possible because  of the underlying control that's fine, I'm pretty sure it should be able to be converted on the consuming side.
- undefined onsubmit label shown on the snackbar after submit, this may be an issue outside of this PR though (p1606177059001400-slack-C01CMSA1VHS)

Otherwise this looks great!

mattsherman committed:
Add transitions when comments field is shown/hidden

mattsherman commented:
>     * undefined onsubmit label shown on the snackbar after submit, this may be an issue outside of this PR though ([p1606177059001400-slack-C01CMSA1VHS](https://href.li/?p1606177059001400-slack-C01CMSA1VHS))

Hmm... I don't see this when submitting... I get the "Thank you for your feedback!" label

mattsherman committed:
Convert score to int before recording

mattsherman commented:
>     * Can the options be numeric instead of strings? If this isn't possible because  of the underlying control that's fine, I'm pretty sure it should be able to be converted on the consuming side.

Since the underlying control is an HTML input of type=radio, the option values need to be strings (I tried changing them to ints, but it does not work properly).

I have added commit d2a62e9 to convert the score to an int before recording it.

mattsherman committed:
Clear error message when score has been selected

mattsherman commented:
>     * The error message ("Please provide feedback by selecting an option above") should be cleared when an option is selected

This has been addressed in e2d0c57

mattsherman commented:
The showing/hidden of the comments field felt a bit jumpy/rough, so I added transitions in 68a7d93

mattsherman committed:
Wrapped comments field UI text for localization

mattsherman commented:
I noticed some comments field-related text that wasn't wrapped for translation 232180d

mattsherman commented:
I have noticed that some of the contrast ratios for foreground/background in this modal are less than 4.5:

<img width="612" alt="Screen Shot 2020-11-24 at 16 25 04" src="https://user-images.githubusercontent.com/2098816/100153544-46912980-2e72-11eb-9b66-cd6254ab67ab.png">

- The selected option labels: 3.8
- The Cancel button: 4.4 (so close!)
- The Send button: 4.4 (so close!)

@elizaan36 Should we be concerned about these? The buttons I believe are using the WP admin theme colors, so that isn't just our issue. But, the selected option label we should probably improve -- do you have any suggestion of colors to change to?

mattsherman commented:
@adrianduffell @becdetat I've pushed a few minor commits to take care of the issues that Bec noted yesterday, as well as a few issues I noticed as well.

I think this is ready for another review (we should really consider merging this assuming there are no major issues found, and addressing outstanding issues in follow-up PRs... this PR has become quite long-lived!).

elizaan36 commented:
> @elizaan36 Should we be concerned about these? The buttons I believe are using the WP admin theme colors, so that isn't just our issue. But, the selected option label we should probably improve -- do you have any suggestion of colors to change to?

Hey @mattsherman, thanks for flagging this. Yep, not much we can do about those buttons at the moment.

One suggestion to fix the color contrast in the modal:
- Can we update the selection text (what appears on the blue background) to Blue 60 (from the Automattic color studio) on hover? The hex code is #135E96.

It would end up looking like this:
<img width="728" alt="Screen Shot 2020-11-24 at 4 34 32 PM" src="https://user-images.githubusercontent.com/5121465/100159204-f4043d00-2e72-11eb-9cc9-5a8b3d7e0128.png">

adrianduffell commented on packages/customer-effort-score/src/customer-feedback-modal/index.js:
```diff
@@ -0,0 +1,175 @@
+		},
+	];
+
+	const [ score, setScore ] = useState( NaN );
```
It doesn't seem right to initialize with `NaN`. I think `NaN` implies there is some sort of error, which isn't the case here.

I think it was better to keep the internal score state as a string, since that is how the radio control natively works and avoids type juggling. I think it is good to convert it to an integer for the `sendScore` parameter though.

becdetat replied:
I don't have a strong opinion either way. I'm happy to roll with this and we can revisit it later if it still feels janky.

mattsherman replied:
Yeah, the use of `NaN` feels a bit too 'clever', but I liked it because it allowed me to not have to specifically account for `null` or `undefined` elsewhere (since `NaN.toString()` works).

In general, I like to keep data in its 'natural' form, and only convert it when needed to fit UI/external integration needs. Score is, at heart, a number. So, I like the idea of treating it as a number whenever possible in our business logic.

But, I don't feel super strongly about this particular case.

becdetat committed:
Increase contrast ratio for focused effort score labels

becdetat commented:
I've just pushed a change to the focused label colour to fix the contrast ratio issue. Once that has gone through I'll merge this.

I can trivially reproduce these issues:

- type error re `CustomerEffortScoreTracks` expecting `trackProps` of type object but received array
- `onSubmitLabel` is undefined

so I'll sort these out in new PRs.

Great work here @adrianduffell , and thanks for your help as well @mattsherman .
